### DISCUSSION TITLE
Runner Set feature request: Add ability to add annotations to Runner Pods once they start running a job

### DISCUSSION THREAD

**Discussion by Nuru** [community/community#82207] [Actions]
### Select Topic Area

Product Feedback: Feature Request

### Body

I made [this request](https://github.com/actions/actions-runner-controller/issues/2562) for the Summerwind ARC and I'd like to make the same request for Runner Sets. 

### What would you like added?

Currently, you can add annotations to every Pod in an AutoscalingRunnerSet by adding them to the `gha-runner-scale-set` Helm chart under 

```
template:
  metadata:
    annotations:
```

I would like the ability to specify annotations to be added to Pods at the time the Pods are assigned jobs, so that idle Pods waiting for jobs do not have the same annotations as Pods running jobs.

### Why is this needed?

Kubernetes cluster autoscaling solutions (e.g. [Karpenter](https://karpenter.sh/)) generally expect that a Pod runs a service that can be terminated on one Node and restarted on another with only a short duration needed to finish processing any in-flight requests. When the cluster is resized, the Cluster Autoscaler will do just that. However, GitHub Action Runner Jobs do not fit this model. If a Pod is terminated in the middle of a job, the job is lost. The likelihood of this happening  is increased by the fact that the Action Runner Controller Autoscaler is expanding and contracting the size of the Runner Pool on a regular basis, causing the Cluster Autoscaler to more frequently want to scale up or scale down the EKS cluster, and, consequently, to move Pods around.

In order to handle situations like this, cluster autoscalers typically allow Pods to indicate, via an annotation, that they cannot be safely interrupted. For the [Kubernetes Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler), you can [add the annotation](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node)

```yaml
"cluster-autoscaler.kubernetes.io/safe-to-evict": "false"
```

For [Karpenter v0.31](https://karpenter.sh/), you can add the [annotation](https://karpenter.sh/v0.31/concepts/deprovisioning/#pod-level-controls)

```yaml
karpenter.sh/do-not-evict: "true"
```

An annotation like this should be added to Pods once they start running jobs, so that the job can finish. 

It needs to be added when a job starts because we do not want this annotation on idle Pods waiting for jobs. Otherwise, the Cluster Autoscaler would be prevented from removing nodes where the idle Pods are waiting, which is exactly the opposite of what we want.

The obvious solution is to have the Runner Set Controller add the annotation to the Pod once a job is assigned to it. In the case of persistent runners, the annotation should be removed once the job is finished.

In order to preserve the zero-permission security of the Runner Pods, these annotations should be managed by the controller, rather than via runner pre- and post-job scripts (which would also require custom runner images with scripts and `kubectl` installed).

### Additional context

It is practically impossible to run very long jobs on a Runner which the Cluster Autoscaler can terminate and evict unless the cluster is very stable in its capacity. Currently the only acceptable solutions is:

- Set `minReplicas =  0` and add the annotation to all pods, solving the problem by never leaving idle Pods deployed

This solution is less desirable because of the lack of idle Runners to pick up jobs quickly.

**Comment by kimxogus** [comment]
I really need this feature to use runner set in karpenter environment
